src/js/node/events.js

```diff
@@ -92,9 +92,28 @@ const emitWithoutRejectionCapture = function emit(type, ...args) {
   if (events === undefined) return false;
   var handlers = events[type];
   if (handlers === undefined) return false;
-
-  for (var handler of [...handlers]) {
-    handler.apply(this, args);
+  // clone handlers array if necessary since handlers can be added/removed during the loop
+  const maybeClonedHandlers = handlers.length > 1 ? handlers.slice() : handlers;
```

Comment:
Instead of `.slice` what about using `arrayClone`:
https://github.com/nodejs/node/blob/18e00a577d7405ca0e6c26113675c230677342c2/lib/events.js#L883-L894

Reply:
I tried that as well but it did not really improve performance. So I just used `slice()` to keep it simple.

#### arrayClone()
```
benchmark                            time (avg)             (min … max)       p75       p99      p995
----------------------------------------------------------------------- -----------------------------
single emit (one handler): bun    46.64 ns/iter  (32.96 ns … 590.09 ns)  38.89 ns 280.46 ns 309.31 ns
single emit (two handlers): bun   71.91 ns/iter  (46.66 ns … 423.27 ns)  68.05 ns  287.8 ns 338.52 ns
on x 10_000 (one handler): bun     5.75 ms/iter    (2.01 ms … 12.03 ms)   7.27 ms  11.81 ms  12.03 ms
on x 10_000 (two handlers): bun    8.48 ms/iter    (3.11 ms … 15.64 ms)   11.2 ms  15.64 ms  15.64 ms
```

Reply:
Ah cool! What if you tested the number of handlers arrayClone optimizes for, so 2-6 (I don't know why 1 is skipped). And then when it breaks over 6 to say the unoptimized 7 handler count.

Reply:
Case 1 is skipped in Node.js because they actually do not store the handlers as an array, if only one handler is attached. But this leads to some ugly `typeof x === 'function' checks 😵.
https://github.com/nodejs/node/blob/main/lib/events.js#L513-L537

Reply:
If Bun's approach is faster or on par and we don't have to unroll as far that's a Win for Bun :)

Reply:
> Case 1 is skipped in Node.js because they actually do not store the handlers as an array, if only one handler is attached. But this leads to some ugly `typeof x === 'function' checks 😵.
> https://github.com/nodejs/node/blob/main/lib/events.js#L513-L537

when i did the initial version of `events.js` i found that JSC was faster when using a single length array vs the function-or-array approach. i forgot if i even tried benching the `arrayClone` function.